### PR TITLE
feat: Implements requirer side of PKI integration

### PR DESCRIPTION
**Pull request by gruyaume**
# Description

This PR implements the requirer side of the PKI interface.
- Upon `certificates-pki-request` relation joined:
  - Configures the PKI mount, configures the PKI intermediate CA (generates a CSR)
  - Sends that CSR over to the relation provider.
- Upon `Certificate Available`:
  - Sets the PKI intermediate CA certificate
  - Sets the PKI charm role

This PR depends on the following PR's being merged first:
- https://github.com/canonical/charm-relation-interfaces/pull/118
- https://github.com/canonical/tls-certificates-interface/pull/83
- https://github.com/canonical/self-signed-certificates-operator/pull/40

To do in further PR's:
- Manage certificate expiring/revoked events
- Implement provider side of the PKI integration

# Checklist:

- [ ] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that validate the behaviour of the software
- [ ] I validated that new and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -530,6 +550,91 @@ def _on_new_vault_kv_client_attached(self, event: NewVaultKvClientAttachedEvent)
...
+        if vault.is_secret_engine_enabled(path=PKI_MOUNT):
+            logger.debug("PKI backend already enabled, skipping")
+            return
+        vault.enable_pki_engine(path=PKI_MOUNT)
+        csr = vault.configure_pki_intermediate_ca(mount=PKI_MOUNT, common_name=common_name)
+        self._pki_requires.request_certificate_creation(
+            certificate_signing_request=csr.encode(), is_ca=True
+        )
+
+    def _on_certificate_pki_request_certificate_available(self, event: CertificateAvailableEvent):
```
Comment (gatici) [inline src/charm.py:587]: What happens if "common_name" is changed after relation_joined ?  Do we need a change in relation in that situation ?

**Review by gatici — CHANGES_REQUESTED** [review]
We need to update the README.md to add the new integration.

File: src/charm.py [hunk 2]
```diff
@@ -530,6 +550,91 @@ def _on_new_vault_kv_client_attached(self, event: NewVaultKvClientAttachedEvent)
...
+            logger.debug("Only leader unit should configure the vault PKI, skipping")
+            return
+        if not self._is_peer_relation_created():
+            logger.debug("Peer relation not created, deferring event")
+            event.defer()
+            return
+        common_name = self._get_config_common_name()
+        if not common_name:
+            logger.error("Common name is not set correctly in the charm configuration")
+            return
```
Comment (ghislainbourgeois) [inline src/charm.py:565]: Seems like we need to handle the configuration change, otherwise the charm will get stuck.

File: src/vault.py [hunk 3]
```diff
@@ -121,6 +121,65 @@ def configure_kv_mount(self, name: str):
                 path=name,
             )
 
+    def enable_pki_engine(self, path: str):
```
Comment (ghislainbourgeois) [inline src/vault.py:124]: I would suggest forcing keyword argument usage for all the methods added:
```suggestion
    def enable_pki_engine(self, *, path: str):
```

**Review by ghislainbourgeois — CHANGES_REQUESTED** [review]

File: src/charm.py (hunk 1 above)
Comment (gruyaume) [reply src/charm.py:611]: Agreed, fixed

**Pull request closed without merging**